core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -590,9 +583,28 @@ public Builder removeShard(ShardRouting shardEntry) {
         }
 
         public IndexShardRoutingTable build() {
+            // don't allow more than one shard copy with same id to be allocated to same node
+            assert distinctNodes() : "more than one shard with same id assigned to same node (shards: " + shards + ")";
             return new IndexShardRoutingTable(shardId, Collections.unmodifiableList(new ArrayList<>(shards)));
         }
 
+        private boolean distinctNodes() {
```

Comment:
If you passed in `shards` here, you could make this static and public, and then unit test it. What do you think?